Add recommendation for selecting HTML id attribute (#25292)
mdn/content: files/en-us/web/html/global_attributes/id/index.md

> **Warning:** This attribute's value is an opaque string: this means that web authors should not rely on it to convey human-readable information (although having your IDs somewhat human-readable can be useful for code comprehension, e.g. consider `ticket-18659` versus `r45tgfe-freds&$@`).

An `id`'s value must not contain {{glossary("whitespace")}} (spaces, tabs etc.). Browsers treat non-conforming IDs that contain whitespace as if the whitespace is part of the ID. In contrast to the [`class`](/en-US/docs/Web/HTML/Global_attributes#class) attribute, which allows space-separated values, elements can only have one single ID value.

Technically, the value for an `id` attribute may contain any character, except {{glossary("whitespace")}} characters. However, to avoid inadvertent errors, only {{glossary("ASCII")}} letters, digits, `'_'`, and `'-'` should be used and the value for an `id` attribute should start with a letter.

> For example, `.` has a special meaning in CSS (it acts as a [class selector](/en-US/docs/Web/CSS/Class_selectors)). Unless you are careful to escape it in the CSS, it won't be recognized as part of the value of an `id` attribute. It is easy to forget to do this, resulting in bugs in your code that could be hard to detect.

Reviewer: ```suggestion
For example, `.` has a special meaning in CSS (it acts as a [class selector](/en-US/docs/Web/CSS/Class_selectors)). Unless you are careful to escape it when used as part of a CSS selector, including as part of a [`querySelector()`](/en-US/docs/Web/API/Document/querySelector) or  [`querySelectorAll()`](/en-US/docs/Web/API/Document/querySelectorAll) parameter, it won't be recognized as part of the value of an `id` attribute. It is easy to forget to do this, resulting in bugs in your code that could be hard to detect.
```

Reviewer: I added this b/c i think it's important to note it's not only about CSS, and this was one way to do it. It actually works with getElementById(). Who knew?

> For example, `.` has a special meaning in CSS (it acts as a [class selector](/en-US/docs/Web/CSS/Class_selectors)). While valid, unless you are careful to escape it when used as part of a CSS selector, including as part of a [`querySelector()`](/en-US/docs/Web/API/Document/querySelector) or  [`querySelectorAll()`](/en-US/docs/Web/API/Document/querySelectorAll) parameter, it won't be recognized as part of the value of an `id` attribute. It is easy to forget to do this, resulting in bugs in your code that could be hard to detect.

Author: ```suggestion
For example, `.` has a special meaning in CSS (it acts as a [class selector](/en-US/docs/Web/CSS/Class_selectors)). While valid, unless you are careful to escape it when used as part of a CSS selector, including as part of a [`querySelector()`](/en-US/docs/Web/API/Document/querySelector) or [`querySelectorAll()`](/en-US/docs/Web/API/Document/querySelectorAll) parameter, it won't be recognized as part of the value of an `id` attribute. It is easy to forget to do this, resulting in bugs in your code that could be hard to detect.
```

> Similarly, an `id` comprised only of digits (E.g., `1234`), or starting with a digit (E.g., `1234-322-678`) or an hyphen (E.g., `-123`), though valid and working well in HTML, may lead to problems when used in CSS, JavaScript and Web APIs:

Reviewer: ```suggestion
Similarly, an `id` starting with a digit (E.g., `1234-322-678`) or a hyphen (E.g., `-123`) followed by a digit, though valid in HTML, may lead to problems when used in CSS, JavaScript, and Web APIs:
```

Reviewer: `comprised only of digits` means it starts with a digit, so removed.

> - CSS selector can only take a CSS identifier, excludint `id` starting with a digit, two hyphens, or one hyphen immediatly followed by a digit: `#544-383-3388` is an invalid selector in CSS and won't work

Reviewer: ```suggestion
- ID selectors can only take a CSS identifier, excluding any `id` starting with a digit, two hyphens, or one hyphen immediately followed by a digit: `#544-383-3388`, `#--varName`, and`-Pi` are invalid selectors and won't work.
```

Reviewer: the two hyphens is working... need to investigate.

Reviewer: removed the double hyphen as it works everywhere.

> - {{domxref("Document.querySelector()")}} and similar methods will not find it (This [page](http://calixe.fr/global_id.html) document the problem).

Reviewer: I added this earlier because i didn't read ahead, so we can get rid of this link which is not `https` and you have to do a view source. 

I built https://codepen.io/estelle/pen/jOvzbgb to test, so we could use that.

Reviewer: I figured out how to escape all of them, so we may want to go with https://codepen.io/estelle/pen/jOvzbgb

Author: What about:
```suggestion
- {{domxref("Document.querySelector()")}} and similar methods will not find them unless you escape them (See this [page](https://codepen.io/estelle/pen/jOvzbgb) for an example.)
```
?

> - Such an `id` will not be a valid JavaScript identifier: `myvar.1234` will not work, while `myvar._1234` would. This may be problematic.

Reviewer 2: I don't think you can access IDs as properties. However, you can access any element with ID as a global variable. This means `document.getElementByID("foo").doSomething()` and `foo.doSomething()` are equivalent. This feature also needs some more documentation so it doesn't come out of the blue. See the answer https://stackoverflow.com/questions/3434278/do-dom-tree-elements-with-ids-become-global-properties

> - Such an `id` will not be a valid JavaScript identifier: `myvar.1234` will not work, while `myvar._1234` would. This may be problematic.

Author: Modified (I think documenting that feature completely is out of scope of this PR)

```suggestion
- Such an `id` will not be valid JavaScript identifiers, as [elements with IDs may become global properties](https://stackoverflow.com/questions/3434278/do-dom-tree-elements-with-ids-become-global-properties): `1234.doSomething()` will not work, whereas `_1234.doSomething()` will. This may be problematic.
```

> **Warning:** This attribute's value is an opaque string: this means that web authors should not rely on it to convey human-readable information (although having your IDs somewhat human-readable can be useful for code comprehension, e.g. consider `ticket-18659` versus `r45tgfe-freds&$@`).

> An `id`'s value must not contain {{glossary("whitespace")}} (spaces, tabs etc.). Browsers treat non-conforming IDs that contain whitespace as if the whitespace is part of the ID. In contrast to the [`class`](/en-US/docs/Web/HTML/Global_attributes#class) attribute, which allows space-separated values, elements can only have one single ID value.

Reviewer: ```suggestion
An `id`'s value must not contain {{glossary("whitespace")}} (spaces, tabs, etc.). Browsers treat non-conforming IDs that contain whitespace as if the whitespace is part of the ID. In contrast to the [`class`](/en-US/docs/Web/HTML/Global_attributes#class) attribute, which allows space-separated values, elements can only have one single ID value.
```

Technically, the value for an `id` attribute may contain any character, except {{glossary("whitespace")}} characters. However, to avoid inadvertent errors, only {{glossary("ASCII")}} letters, digits, `'_'`, and `'-'` should be used and the value for an `id` attribute should start with a letter.

> For example, `.` has a special meaning in CSS (it acts as a [class selector](/en-US/docs/Web/CSS/Class_selectors)). While valid, unless you are careful to escape it when used as part of a CSS selector, including as part of a [`querySelector()`](/en-US/docs/Web/API/Document/querySelector) or [`querySelectorAll()`](/en-US/docs/Web/API/Document/querySelectorAll) parameter, it won't be recognized as part of the value of an `id` attribute. It is easy to forget to do this, resulting in bugs in your code that could be hard to detect.

Reviewer 2: ```suggestion
For example, `.` has a special meaning in CSS (it starts a [class selector](/en-US/docs/Web/CSS/Class_selectors)). While valid, unless you are careful to escape it when used as part of a CSS selector, it won't be recognized as part of the element's `id`. The same applies for the [`querySelector()`](/en-US/docs/Web/API/Document/querySelector) and [`querySelectorAll()`](/en-US/docs/Web/API/Document/querySelectorAll) parameter, which uses the same selector syntax. It is easy to forget to do this, resulting in bugs in your code that could be hard to detect.
```

Similarly, an `id` starting with a digit (E.g., `1234-322-678`) or a hyphen followed by a digit (E.g., `-123`), though valid in HTML, may lead to problems when used in CSS, JavaScript, and Web APIs:

> - CSS [ID selectors](/en-US/docs/Web/CSS/ID_selectors) accept any CSS identifier. If the `id` starts with a digit or one hyphen immediately followed by a digit, the hyphen and digit must be escaped in CSS. For example, while `id="544-383-3388"` and `id="-3Pi"` are valid in HTML, the `id` selectors must be escaped. The element with these `id` values can be targeted in CSS with `#\35 44-383-3388" and `#\2D \33 pi". 

Author: ```suggestion
- CSS [ID selectors](/en-US/docs/Web/CSS/ID_selectors) accept any CSS identifier. If the `id` starts with a digit or one hyphen immediately followed by a digit, the hyphen and digit must be escaped in CSS. For example, while `id="544-383-3388"` and `id="-3Pi"` are valid in HTML, the `id` selectors must be escaped. The element with these `id` values can be targeted in CSS with `#\35 44-383-3388` and `#\2D \33 pi`.
```

- Any valid HTML `id` value is valid as an attribute selector in CSS and JavaScript. The selectors `[id="544-383-3388"]` and `[id="-3Pi"]` are valid.
- {{domxref("Document.querySelector()")}} and similar methods will not find them unless you escape them (See this [page](https://codepen.io/estelle/pen/jOvzbgb) for an example.)

> - Such an `id` will not be valid JavaScript identifiers, as [elements with IDs may become global properties](https://stackoverflow.com/questions/3434278/do-dom-tree-elements-with-ids-become-global-properties): `1234.doSomething()` will not work, whereas `_1234.doSomething()` will. This may be problematic.

Reviewer 2: I will come back in a future PR to add this information to this page. It deserves more than an SO link...

```suggestion
- Such an `id` will not be valid JavaScript identifiers, as [elements with IDs become global properties](https://stackoverflow.com/questions/3434278/do-dom-tree-elements-with-ids-become-global-properties): `1234.doSomething()` will not work, whereas `_1234.doSomething()` will. This is slightly inconvenient as you have to create the variable with an extra step: `const element = document.getElementById("1234")`.
```

Author: Agreed on revisiting this in the future.

> - {{domxref("Document.querySelector()")}} and similar methods will not find them unless you escape them (See this [page](https://codepen.io/estelle/pen/jOvzbgb) for an example.)
> - Such an `id` will not be valid JavaScript identifiers, as [elements with IDs become global properties](https://stackoverflow.com/questions/3434278/do-dom-tree-elements-with-ids-become-global-properties): `1234.doSomething()` will not work, whereas `_1234.doSomething()` will. This is slightly inconvenient as you have to create the variable with an extra step: `const element = document.getElementById("1234")`.

Reviewer 2: ```suggestion
- {{domxref("Document.querySelector()")}} and similar methods using CSS selector style queries will not find them unless you escape them (See this [page](https://codepen.io/estelle/pen/jOvzbgb) for an example.)
- Such an `id` is not a valid JavaScript identifier. [Elements with IDs become global properties](https://stackoverflow.com/questions/3434278/do-dom-tree-elements-with-ids-become-global-properties), but you cannot use non-identifier global properties as global variables — `1234` is not a global variable, and you have to use `window[1234]` instead to get the element with `id="1234"`. This is slightly inconvenient as you have to create the variable with an extra step: `const element = window[1234]`.
```

> Technically, the value for an `id` attribute may contain any character, except {{glossary("whitespace")}} characters. However, to avoid inadvertent errors, only {{glossary("ASCII")}} letters, digits, `'_'`, and `'-'` should be used and the value for an `id` attribute should start with a letter.
> 
> For example, `.` has a special meaning in CSS (it starts a [class selector](/en-US/docs/Web/CSS/Class_selectors)). While valid, unless you are careful to escape it when used as part of a CSS selector, it won't be recognized as part of the element's `id`. The same applies to the [`querySelector()`](/en-US/docs/Web/API/Document/querySelector) and [`querySelectorAll()`](/en-US/docs/Web/API/Document/querySelectorAll) parameters, which use the same selector syntax. It is easy to forget to do this, resulting in bugs in your code that could be hard to detect.
> 
> Similarly, an `id` starting with a digit (E.g., `1234-322-678`) or a hyphen followed by a digit (E.g., `-123`), though valid in HTML, may lead to problems when used in CSS, JavaScript, and Web APIs:
> 
> - CSS [ID selectors](/en-US/docs/Web/CSS/ID_selectors) accept any CSS identifier. If the `id` starts with a digit or one hyphen immediately followed by a digit, the hyphen and digit must be escaped in CSS. For example, while `id="544-383-3388"` and `id="-3Pi"` are valid in HTML, the `id` selectors must be escaped. The element with these `id` values can be targeted in CSS with `#\35 44-383-3388` and `#\2D \33 pi`.
> - Any valid HTML `id` value is valid as an attribute selector in CSS and JavaScript. The selectors `[id="544-383-3388"]` and `[id="-3Pi"]` are valid.
> - {{domxref("Document.querySelector()")}} and similar methods using CSS-selector-style queries will not find them unless you escape them. (See this [page](https://codepen.io/estelle/pen/jOvzbgb) for an example.)
> - Such an `id` is not a valid JavaScript identifier. [Elements with IDs become global properties](https://stackoverflow.com/questions/3434278/do-dom-tree-elements-with-ids-become-global-properties), but you cannot use non-identifier global properties as global variables — `1234` is not a global variable, and you have to use `window[1234]` instead to get the element with `id="1234"`. This is slightly inconvenient as you have to create the variable with an extra step: `const element = window[1234]`.

Reviewer: ```suggestion
Technically, the value for an `id` attribute may contain any character, except {{glossary("whitespace")}} characters. However, to avoid inadvertent errors, only {{glossary("ASCII")}} letters, digits, `'_'`, and `'-'` should be used, and the value for an `id` attribute should start with a letter.

For example, `.` has a special meaning in CSS (it starts a [class selector](/en-US/docs/Web/CSS/Class_selectors)). While valid, unless you are careful to escape it when used as part of a CSS selector, it won't be recognized as part of the element's `id`. The same applies to the [`querySelector()`](/en-US/docs/Web/API/Document/querySelector) and [`querySelectorAll()`](/en-US/docs/Web/API/Document/querySelectorAll) parameters, which use the same selector syntax. It is easy to forget to do this, resulting in bugs in your code that could be hard to detect.

Similarly, an `id` starting with a digit (E.g., `1234-322-678`) or a hyphen followed by a digit (E.g., `-123`), though valid in HTML, may lead to problems when used in CSS, JavaScript, and Web APIs:

- CSS [ID selectors](/en-US/docs/Web/CSS/ID_selectors) accept any CSS identifier. If the `id` starts with a digit or one hyphen immediately followed by a digit, both the hyphen and digit must be escaped in CSS. For example, while `id="544-383-3388"` and `id="-3Pi"` are valid in HTML, the `id` selectors must be escaped. The element with these `id` values can be targeted in CSS with `#\35 44-383-3388` and `#\2D \33 pi`.
- Any valid HTML `id` value is valid as an attribute selector in CSS and JavaScript. The selectors `[id="544-383-3388"]` and `[id="-3Pi"]` are valid.
- {{domxref("Document.querySelector()")}} and similar methods using CSS-selector-style queries will not find them unless you escape them. (See this [page](https://codepen.io/estelle/pen/jOvzbgb) for an example.)
- Such an `id` is not a valid JavaScript identifier. [Elements with IDs become global properties](https://stackoverflow.com/questions/3434278/do-dom-tree-elements-with-ids-become-global-properties), but you cannot use non-identifier global properties as global variables — `1234` is not a global variable, and you have to use `window[1234]` instead to get the element with `id="1234"`. This is slightly inconvenient as you have to create the variable with an extra step: `const element = window[1234]`.
```
